lib/src/socket_connector.dart

```diff
@@ -219,16 +233,24 @@ class SocketConnector {
         close();
       }
     }
-    side.state = SideState.closed;
+
     try {
       _log(chalk.brightBlue('Destroying socket on side ${side.name}'));
+      await side.socket.flush();
       side.socket.destroy();
-      if (side.farSide != null) {
-        _log(chalk.brightBlue(
-            'Destroying socket on far side (${side.farSide?.name})'));
-        _destroySide(side.farSide!);
+      if (side.farSide != null && side.farSide!.state != SideState.closed) {
+        if (side.rcvd == side.farSide!.sent) {
+          _log(chalk.brightBlue(
+              'Far side (${side.farSide?.name}) has received all data - will close it'));
+          _closeSide(side.farSide!);
+        } else {
+          _log(chalk.brightBlue(
+              'Far side (${side.farSide?.name}) has NOT YET received all data'));
+        }
```

Comment:
Check that all data received from this side has been delivered to the far side before calling _closeSide(farSide). Since sometimes, because of this check, _closeSide will not be called here, blocks of code in the listen onData functions above perform the same check when writing data to the far side's sink; the combination of all of this guarantees that all data is always written to the other side before closing, and guarantees that the sides are always closed (and thus their sockets flushed and destroyed) once all data has been written.